Added a note about hardware requirements for compiling on Linux

art-divin opened the pull request:
<!-- What's in this pull request? -->

When I was setting up a Linux VM environment for debugging Swift compiler crash, I faced a number of issues due to low disk space (`update-checkout` would hang before starting concurrent checkouts) and due to low RAM size (linker would start failing when linking symbols for `bin/clang-17` and other binaries).

This PR adds a note about the minimal RAM requirements, and updates the recommended minimum available disk space.

My findings for when running a Linux VM and compiling the compiler are as follows:

1. 8 GB of RAM with `--release-debuginfo` flag
2. 32 GB of RAM with `--debug` flag
3. 150 GB of disk space

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/main/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

finagolfin commented on docs/HowToGuides/GettingStarted.md:
```diff
@@ -190,6 +190,10 @@ toolchain as a one-off, there are a couple of differences:
 > with version 2 shipped with Ubuntu.
 
+> **Note**  
+> When running Linux in a VM, make sure to have at least 8 GB (when using flag `--release-debuginfo`) and 32 GB (when using flag `--debug-swift`) of RAM and 150 GB of disk space available. To speed up compilation of the toolchain, it is better to dedicate as many CPUs as possible on the host machine.
```
"running Linux in a VM" -> "building on linux" as the same requirements outside VM
"8 GB (when using flag `--release-debuginfo`)" -> "8 GB of RAM when using flag `--release-debuginfo`"
"32 GB (when using flag `--debug-swift`) of RAM and 150 GB" -> "32 GB of RAM when using flag `--debug-swift`, plus 150 GB"

I think you can cut the last two sentences you added, starting with "To speed up."

art-divin replied:
👋🏻  Hello @finagolfin ,

thank you very much for your input (and posts on Swift Forums!). Really essential, always.

Updated the text 👍🏻

finagolfin commented on docs/HowToGuides/GettingStarted.md:
```diff
@@ -190,6 +190,9 @@ toolchain as a one-off, there are a couple of differences:
 > with version 2 shipped with Ubuntu.
 
+> **Note**  
+> When building on Linux, make sure to have at least 8 GB when using flag `--release-debuginfo` and 32 GB when using flag `--debug-swift` of RAM and 150 GB of disk space available.
```
While it may be a bit redundant, it reads better if you write "GB of RAM" twice, rather than tuck that away at the end.

art-divin replied:
Thank you, improved! 🔥

finagolfin reviewed (approved):
Other than nit on phrasing, seems worthwhile.

finagolfin commented on docs/HowToGuides/GettingStarted.md:
```diff
@@ -190,6 +190,9 @@ toolchain as a one-off, there are a couple of differences:
 > with version 2 shipped with Ubuntu.
 
+> **Note**  
+> When building on Linux, make sure to have at least 8 GB of RAM when using flag `--release-debuginfo` and 32 GB of RAM when using flag `--debug-swift` and 150 GB of disk space available.
```
and 150 -> , plus 150

Too many `and`s otherwise

art-divin replied:
Adjusted 🤝

AnthonyLatsis commented on docs/HowToGuides/GettingStarted.md:
```diff
@@ -190,6 +190,9 @@ toolchain as a one-off, there are a couple of differences:
 > LLDB currently requires at least `swig-1.3.40` but will successfully build
 > with version 2 shipped with Ubuntu.
 
+> **Note**  
+> When building on Linux, make sure to have at least 8 GB of RAM when using flag `--release-debuginfo` and 32 GB of RAM when using flag `--debug-swift`, plus 150 GB of disk space available.
+
```
* I think this should go under "System Requirements". Perhaps we should update the default disk space requirement to 100GB too.
* What’s taking up 150GB on Linux? The entire checkout plus a `--release-debuginfo --debug-swift` build is only about 70GB for me on macOS.
* How much RAM was not enough to pull through a `--debug-swift` build, 16GB?

art-divin replied:
> I think this should go under "System Requirements". Perhaps we should update the default disk space requirement to 100GB too.

Amazing, I did not even notice this section! Surely, updated 🤝 

> What’s taking up 150GB on Linux? The entire checkout plus a --release-debuginfo --debug-swift build is only about 70GB for me on macOS.

--debug-swift took 78 GB of disk space, plus according to [Development Tips](https://github.com/apple/swift/blob/main/docs/DevelopmentTips.md) it is recommended to set sccache `SCCACHE_CACHE_SIZE` environment variable to 50 GB. Then, system size is about 19 GB. 
I've ran out of space mostly because I am still unable to compile with `--debug-swift --debug-foundation` flags, and have to resort to `--release-debuginfo -T --debug-swift-stdlib` to make some progress in crash investigation. Nonetheless, 150 GB is enough to avoid running into a problem with low disk space in the middle of LLVM compilation, for example.

|Folders|Size|
:-:|:-:
<img width="427" alt="image" src="https://github.com/apple/swift/assets/1614869/a57f7195-5cd9-4377-b462-040ec6d00e61">|<img width="253" alt="image" src="https://github.com/apple/swift/assets/1614869/d1485575-78e1-42c5-a26d-fea1cf274a9d">

> How much RAM was not enough to pull through a --debug-swift build, 16GB?
Answering your question, 4 GB was my initial VM setup, and I wasted 3 days figuring out why linking was failing, discovered the fact by accident. So my intention with this PR is to let others know how to avoid redundant investigations and work.

Even with 32 GB of RAM, linking of `clang-17` binary fails every single time in a VM. On hardware Linux, it'd be better I presume; however, after 2-3 tries I am able to link largest binaries. On the other hand, if to limit concurrent tasks with `-j 5`, for instance, failures are more rare, but still occur for the same reason, not enough RAM available.

AnthonyLatsis reviewed (approved):
Thank you! Just one more suggestion.

art-divin commented on docs/HowToGuides/GettingStarted.md:
```diff
@@ -53,9 +53,14 @@ toolchain as a one-off, there are a couple of differences:
-5. Time:
+   Build artifacts take anywhere between 5 GB to 100 GB, depending on the
+   build settings. It is recommended to have at least 150 GB of available disk space.
+5. RAM:
+   On Linux, it is recommended to have at least 8 GB of RAM when compiling
+   with the `--release-debuginfo` build flag and at least 16 GB of RAM available
+   when compiling with the `--debug-swift` build flag. 
+   For the best experience, 32 GB of RAM is recommended.
```
Very nice rephrasing, thank you for your input 🤝

art-divin committed:
Added a note about RAM and disk space hardware requirements

commit bcc59744b7da00395708f41984ee2be70924cb6c
Author: Ruslan Alikhamov <r.alikhamov@gmail.com>
Date:   Sat Jan 13 15:28:42 2024 +0400

    applied suggested phrasing

commit b5b0f9622a5dca80cf79df41e4834959021329ca
Author: Ruslan Alikhamov <r.alikhamov@gmail.com>
Date:   Fri Jan 12 23:48:12 2024 +0400

    Added a note about hardware requirements for compiling on Linux

    commit 863c1d98dd8c3c3e67610fe9ea9ac00e161d687a
    Author: Ruslan Alikhamov <r.alikhamov@gmail.com>
    Date:   Fri Jan 12 23:47:19 2024 +0400

        Moved note to Requirements section

    commit 307d0027967945cdc5c0d28b9c4525fc27149f66
    Author: Ruslan Alikhamov <r.alikhamov@gmail.com>
    Date:   Wed Jan 10 19:45:54 2024 +0400

        Added a note about hardware requirements for compiling on Linux

        commit 943ccc2726aa2f8850f38670b848304acb3359aa
        Author: Ruslan Alikhamov <r.alikhamov@gmail.com>
        Date:   Wed Jan 10 19:45:13 2024 +0400

            Replaced 'and' with 'plus' to ease comprehension

        commit 3427ace37d63d398542aa80be6c30a27a940eef9
        Author: Ruslan Alikhamov <r.alikhamov@gmail.com>
        Date:   Thu Jan 4 21:02:22 2024 +0400

            Added a note about hardware requirements for compiling on Linux

            commit 49bafe98a3c938e83af3de351094b47650836aee
            Author: Ruslan Alikhamov <r.alikhamov@gmail.com>
            Date:   Thu Jan 4 20:49:35 2024 +0400

                Added "of RAM" to each reference

            commit 9dc879cb30e9c9507d78d5de3a1d0de2808979c6
            Author: Ruslan Alikhamov <r.alikhamov@gmail.com>
            Date:   Thu Jan 4 14:13:20 2024 +0400

                Updated text according to comment, removed mention of VM for Linux

            commit b652195dcbedc7fa3a770075457880d544a78a60
            Author: Ruslan Alikhamov <r.alikhamov@gmail.com>
            Date:   Wed Jan 3 00:06:18 2024 +0400

                Added more information about different RAM requirements

            commit 2aa1be00ce6ab66f50928484727e684a9eb5c243
            Author: Ruslan Alikhamov <r.alikhamov@gmail.com>
            Date:   Mon Jan 1 03:19:22 2024 +0400

                Added a note about Linux inside of a VM hardware requirements

AnthonyLatsis commented:
@xedin It’s your call now!

art-divin commented:
Thank you very much for your approvals 🤝 

I wonder what would be the next step, shall we run CI for such change, or just merge? How to proceed, I'd appreciate any hint 👍🏻

AnthonyLatsis commented:
We’re waiting for a code owner to also approve your changes.

ping @xedin

AnthonyLatsis commented:
@swift-ci please smoke test

AnthonyLatsis commented:
Thanks!
